File: ca-gpt-free-v9/shared/permissions.py
# This file is the single source of truth for all permissions in the system.
# It is shared between the Python services.

from typing import Final, List, Literal

class Permissions:
    # Admin & User Management
    MANAGE_USERS: Final = 'manage_users'
    MANAGE_ROLES: Final = 'manage_roles'
    VIEW_ADMIN_PANEL: Final = 'view_admin_panel'

    # System Admin
    MANAGE_PROMPTS: Final = 'manage_prompts'

    # Audit
    VIEW_AUDIT_LOGS: Final = 'view_audit_logs'

    # Accounting
    CREATE_JOURNAL_ENTRY: Final = 'create_journal_entry'
    VIEW_JOURNAL: Final = 'view_journal'
    VIEW_LEDGER: Final = 'view_ledger'
    VIEW_TRIAL_BALANCE: Final = 'view_trial_balance'
    GENERATE_PL_REPORT: Final = 'generate_pl_report'
    GENERATE_BS_REPORT: Final = 'generate_bs_report'

    # GST
    FILE_GST_RETURN: Final = 'file_gst_return'
    VIEW_GST_REPORTS: Final = 'view_gst_reports'

    # Direct Tax
    COMPUTE_TAX: Final = 'compute_tax'
    FILE_ITR: Final = 'file_itr'

    # Payroll
    RUN_PAYROLL: Final = 'run_payroll'
    VIEW_PAYROLL_REPORTS: Final = 'view_payroll_reports'

    # ROC
    FILE_ROC: Final = 'file_roc'
    VIEW_ROC_FILINGS: Final = 'view_roc_filings'

    @classmethod
    def get_all(cls) -> List[str]:
        return [
            value for name, value in vars(cls).items()
            if not name.startswith("_") and isinstance(value, str)
        ]

Permission = Literal[
    'manage_users', 'manage_roles', 'view_admin_panel', 'manage_prompts',
    'view_audit_logs', 'create_journal_entry', 'view_journal', 'view_ledger',
    'view_trial_balance', 'generate_pl_report', 'generate_bs_report',
    'file_gst_return', 'view_gst_reports', 'compute_tax', 'file_itr',
    'run_payroll', 'view_payroll_reports', 'file_roc', 'view_roc_filings'
]

ALL_PERMISSIONS: List[Permission] = Permissions.get_all()

File: ca-gpt-free-v9/shared/permissions.ts
// This file is the single source of truth for all permissions in the system.
// It is shared between the frontend and backend.

export const Permissions = {
  // Admin & User Management
  MANAGE_USERS: 'manage_users', // Create, update, delete users
  MANAGE_ROLES: 'manage_roles', // Create, update, delete roles and their permissions
  VIEW_ADMIN_PANEL: 'view_admin_panel',

  // System Admin
  MANAGE_PROMPTS: 'manage_prompts', // Edit AI prompts for the company

  // Audit
  VIEW_AUDIT_LOGS: 'view_audit_logs',

  // Accounting
  CREATE_JOURNAL_ENTRY: 'create_journal_entry',
  VIEW_JOURNAL: 'view_journal',
  VIEW_LEDGER: 'view_ledger',
  VIEW_TRIAL_BALANCE: 'view_trial_balance',
  GENERATE_PL_REPORT: 'generate_pl_report',
  GENERATE_BS_REPORT: 'generate_bs_report',

  // GST
  FILE_GST_RETURN: 'file_gst_return',
  VIEW_GST_REPORTS: 'view_gst_reports',

  // Direct Tax
  COMPUTE_TAX: 'compute_tax',
  FILE_ITR: 'file_itr',

  // Payroll
  RUN_PAYROLL: 'run_payroll',
  VIEW_PAYROLL_REPORTS: 'view_payroll_reports',

  // ROC
  FILE_ROC: 'file_roc',
  VIEW_ROC_FILINGS: 'view_roc_filings',
} as const;

export type Permission = (typeof Permissions)[keyof typeof Permissions];

export const ALL_PERMISSIONS = Object.values(Permissions) as Permission[];

export const DefaultRoles = {
  Admin: {
    name: 'Admin',
    description: 'Has all permissions and can manage the system.',
    permissions: ALL_PERMISSIONS,
  },
  ElevatedUser: {
    name: 'ElevatedUser',
    description: 'Can perform most financial tasks but cannot manage users or system settings.',
    permissions: [
      Permissions.CREATE_JOURNAL_ENTRY,
      Permissions.VIEW_JOURNAL,
      Permissions.VIEW_LEDGER,
      Permissions.VIEW_TRIAL_BALANCE,
      Permissions.GENERATE_PL_REPORT,
      Permissions.GENERATE_BS_REPORT,
      Permissions.FILE_GST_RETURN,
      Permissions.VIEW_GST_REPORTS,
      Permissions.COMPUTE_TAX,
      Permissions.FILE_ITR,
      Permissions.RUN_PAYROLL,
      Permissions.VIEW_PAYROLL_REPORTS,
      Permissions.FILE_ROC,
      Permissions.VIEW_ROC_FILINGS,
    ],
  },
  NormalUser: {
    name: 'NormalUser',
    description: 'Can view reports and perform basic data entry.',
    permissions: [
      Permissions.CREATE_JOURNAL_ENTRY,
      Permissions.VIEW_JOURNAL,
      Permissions.VIEW_LEDGER,
      Permissions.VIEW_TRIAL_BALANCE,
      Permissions.GENERATE_PL_REPORT,
      Permissions.GENERATE_BS_REPORT,
      Permissions.VIEW_GST_REPORTS,
      Permissions.VIEW_PAYROLL_REPORTS,
      Permissions.VIEW_ROC_FILINGS,
    ],
  },
  ReadOnlyUser: {
    name: 'ReadOnlyUser',
    description: 'Can only view data and reports.',
    permissions: [
        Permissions.VIEW_JOURNAL,
        Permissions.VIEW_LEDGER,
        Permissions.VIEW_TRIAL_BALANCE,
        Permissions.GENERATE_PL_REPORT,
        Permissions.GENERATE_BS_REPORT,
        Permissions.VIEW_GST_REPORTS,
        Permissions.VIEW_PAYROLL_REPORTS,
        Permissions.VIEW_ROC_FILINGS,
    ],
  }
};